### ISSUE TITLE
[Bug]: Error: Invalid index when accessing data.linode_images

### ISSUE THREAD

**Issue by ajprok** [linode/terraform-provider-linode#1033]
### Terraform Version

Terraform v1.5.7 on linux_amd64

### Linode Provider Version

version = "2.7.2"

### Effected Terraform Resources

linode_images

### Directory structure
```
├── main.tf
└── submodule
    └── main.tf
```
### main.tf file contents
```
terraform {
  required_providers {
    linode = {
      source  = "linode/linode"
      version = "2.7.2"
    }
  }
}
provider "linode" {
  token = "<api-token>"
}
resource "linode_sshkey" "main" {
  label   = "ansible"
  ssh_key = chomp(file("~/.ssh/ansible.pub"))
}
module "submodule" {
  source     = "./submodule"
  depends_on = [linode_sshkey.main]
}
output "image_id" {
  value = module.submodule.image_id
}
```
### submodule/main.tf file contents
```
terraform {
  required_providers {
    linode = {
      source = "linode/linode"
    }
  }
}
data "linode_images" "main" {
  latest = true
  filter {
    name   = "label"
    values = ["<label_of_image_which_already_exists"]
  }
}
output "image_id" {
  value = data.linode_images.main.images.0.id
}
```
### Terraform plan output
```
Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create
 <= read (data resources)

Terraform planned the following actions, but then encountered a problem:

  # linode_sshkey.main will be created
  + resource "linode_sshkey" "main" {
      + created = (known after apply)
      + id      = (known after apply)
      + label   = "ansible"
      + ssh_key = "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDc5xvCOkpOBgCxaBaK2B5kp8HkmYEGTOoJVOE+UNuxy78M2H2n13e6Qti+IV2jc1K85VNwxCDKpLHfwe3DXy9aKSrn2XfviqMH9wt5dE/wmmh8i3cL4S8n698PEOwUr0o2G5Dj7559FF6gglbm6yEOwfHpZv2CFEHsh8AogfPj+uj3mRjkVJuJLbzeO42nwUzxbtx4npc+W/jfEmbdXDARkTbELW3Mj0DcC8mhbJTzPDLxQVlFH+9yHnz5ah8riaNqrT2zeysVVBeQFxyg0SlOVLYnXhSPTkRyIxXBTfZ2N1zWq/gKjl8vvzIhz8u9CWXnhN15bJemqqMmG4rXgQ6uVaIT+VpbwTjkc1XFDIznM4B8xPkkP57iezAs88itKHk3AqqZ8Wzgb+hRz/W3qYkVyl1kfOG7Ab9VgCUKmicCzyHz5tK3akC4AEptB4Q7T31o9i/3WbMjWLrhpPJLJKFimhr7jppBAbClEE8kGU1XsaghgTT7rB6cWrJQrHpXKAjn9ivZngPCVBI1Brh8KWSk5YX/MB/GaLWna/SpooFyrsgjmVkUROsRroGAbrZ7kRl+vwvT1CSEEJawjB9BGeZaiBE+FaUQIZA3Q2blO88pKs3g13xGbGFNayRmol9APwN570AlRsZi1yT6PmSfqjXjZPDMnGVJxSDgba65qchLVw== ansible"
    }

  # module.submodule.data.linode_images.main will be read during apply
  # (depends on a resource or a module with changes pending)
 <= data "linode_images" "main" {
      + id     = (known after apply)
      + latest = true

      + filter {
          + name   = "label"
          + values = [
              + "<label_of_image_which_already_exists",
            ]
        }
    }

Plan: 1 to add, 0 to change, 0 to destroy.
╷
│ Error: Invalid index
│ 
│   on submodule/main.tf line 19, in output "image_id":
│   19:   value = data.linode_images.main.images.0.id
│     ├────────────────
│     │ data.linode_images.main.images is empty list of object
│ 
│ The given key does not identify an element in this collection value: the collection has no elements.
```

### Expected Behavior

This is an interesting bug. When the submodule which calls the **data.linode_images** has a **depends_on** entry, we get the `Error: Invalid index`. Otherwise, it works properly.

**Expected behaviour**: This setup should work properly even when **depends_on** entry is present.

_Workaround hack for people who are stuck:
Create a local variable which creates a dependency for **data.linode_images.main.images** on **data.linode_images.main.id**. This makes data.linode_images.main.images a **known after apply** property, and the error is by passed.
Example_
```
locals {
  image_id = length(data.linode_images.main.id) > 0 ? data.linode_images.main.images.0.id : null
}
output "image_id" {
  value = local.image_id
}
```

### Actual Behavior

We get `Error: Invalid index`.

### Steps to Reproduce

Make sure that you are performing a **fresh** terraform plan. Else you can also first do a terraform destroy & then a terraform apply.

- terraform init
- terraform plan

**Comment by ajprok** [comment]
There may be a connection with the linode_images schema. This [link](https://discuss.hashicorp.com/t/terraform-0-13-handling-of-data-source-data-resource-reads-can-no-longer-be-disabled-by-refresh-false/18457/8) may help in providing perspective.

**Comment by yec-akamai** [comment]
Hi @ajprok, thanks for reporting this issue! 

After looking into it, I think it's the way that how `depends_on` works interrupts the read of the data source `linode_images`. When you use `depends_on`, it lets Terraform to finish all actions on the dependency (`linode_sshkey.main`) before performing any actions on the submodule object. That's why the image list is empty. In this case, without explicitly specifying the dependency should address your issue.

Let me know if it helps! You can also find more information about the usage of `depends_on` here: https://developer.hashicorp.com/terraform/language/meta-arguments/depends_on

**Comment by ajprok** [comment]
> Hi @ajprok, thanks for reporting this issue!
> 
> After looking into it, I think it's the way that how `depends_on` works interrupts the read of the data source `linode_images`. When you use `depends_on`, it lets Terraform to finish all actions on the dependency (`linode_sshkey.main`) before performing any actions on the submodule object. That's why the image list is empty. In this case, without explicitly specifying the dependency should address your issue.
> 
> Let me know if it helps! You can also find more information about the usage of `depends_on` here: https://developer.hashicorp.com/terraform/language/meta-arguments/depends_on

Thank you for the information. Unfortunately, in my situation, I require depends_on. The terraform example, I provided earlier was just a simple version that showcases the bug. In my opinion, the images property that is provided by the  linode_images schema should be a computed value. The link I provided earlier may help you diagnose further.

**Comment by yec-akamai** [comment]
> > Hi @ajprok, thanks for reporting this issue!
> > After looking into it, I think it's the way that how `depends_on` works interrupts the read of the data source `linode_images`. When you use `depends_on`, it lets Terraform to finish all actions on the dependency (`linode_sshkey.main`) before performing any actions on the submodule object. That's why the image list is empty. In this case, without explicitly specifying the dependency should address your issue.
> > Let me know if it helps! You can also find more information about the usage of `depends_on` here: https://developer.hashicorp.com/terraform/language/meta-arguments/depends_on
> 
> Thank you for the information. Unfortunately, in my situation, I require depends_on. The terraform example, I provided earlier was just a simple version that showcases the bug. In my opinion, the images property that is provided by the linode_images schema should be a computed value. The link I provided earlier may help you diagnose further.

The `images` block is computed. However, the `depends_on` defers the read of `linode_images` data source to `apply`, because there are unknown (known after apply) values in the `linode_sshkey` resource (id and created). It's a Terraform behavior which we don't have much to control on our side:  
```
  # module.submodule.data.linode_images.main will be read during apply
  # (depends on a resource or a module with changes pending)
```

May I get more context to understand why `depends_on` is required in your use case? We can look into other solutions. I also find building custom conditions can work in this case if you're interested:
```
resource "linode_sshkey" "main" {
  label   = "ansible"
  ssh_key = chomp(file("~/.ssh/ansible.pub"))
  lifecycle {
    postcondition {
      condition     = length(module.submodule.image_id) > 0
      error_message = "Failed to read images."
    }
  }
}
```

**Comment by ajprok** [comment]
In my situation, ssh keys resources are created first. Only after that is created, using depends_on, my submodule is created. In the submodule, the ssh keys are accessed through data block & then passed to the instance resource. In this submodule is also where I access the image through the data block.

**Comment by yec-akamai** [comment]
Hi @ajprok, sorry for the delay. Unfortunately we don't think `depends_on` can work for your use case, because the explicit dependency will always defer the read of the image data source in the submodule until apply. 

One workaround that we can think of is using implicit dependency, i.e. https://gist.github.com/lgarber-akamai/289e8ea1483dad2f37fffde6b75d8d5b (cc @lgarber-akamai )

Another workaround is considering moving the read of the images data source out of the submodule if possible. It sounds like that the instance resource is the actual one in the submodule that depends on the ssh keys creation. If you can move the read of the images to another place, it may not be deferred by the ssh creation in the plan stage.
